numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3981,6 +3981,9 @@ array_shares_memory_impl(PyObject *args, PyObject *kwds, Py_ssize_t default_max_
     }
     else if (PyLong_Check(max_work_obj)) {
         max_work = PyLong_AsSsize_t(max_work_obj);
+        if (PyErr_Occurred()) {
```

Comment:
The Python 3 manual doesn't mention the -1 return value, so I didn't add that...